### ISSUE TITLE
CodeReady Studio 'crc setup detection' algorithm broken with CRC 1.24.0

### ISSUE THREAD

**Issue by jeffmaury** [code-ready/crc#2170]
### General information

  * OS: Windows
  * Hypervisor: Hyper-V
  * Did you run `crc setup` before starting it (Yes/No)?No
  * Running CRC on: Laptop

## CRC version
```bash
CodeReady Containers version: 1.24.0+5f06e84b
OpenShift version: 4.7.2 (embedded in executable)
```
  
## CRC status
```bash
# Put `crc status --log-level debug` output here
```

## CRC config
```bash
Machine does not exist. Use 'crc start' to create it
```

## Host Operating System
```bash

Nom de l’hôte:                              DESKTOP-U2EVGBM
Nom du système d’exploitation:              Microsoft Windows 10 Professionnel
Version du système:                         10.0.19042 N/A build 19042
Fabricant du système d’exploitation:        Microsoft Corporation
Configuration du système d’exploitation:    Station de travail autonome
Type de build du système d’exploitation:    Multiprocessor Free
Propriétaire enregistré:                    Jeff MAURY
Organisation enregistrée:                   
Identificateur de produit:                  00330-80819-37372-AA911
Date d’installation originale:              29/06/2020, 14:54:33
Heure de démarrage du système:              29/03/2021, 15:14:51
Fabricant du système:                       LENOVO
Modèle du système:                          20LAS3NJ0H
Type du système:                            x64-based PC
Processeur(s):                              1 processeur(s) installé(s).
                                            [01] : Intel64 Family 6 Model 142 Stepping 10 GenuineIntel ~1910 MHz
Version du BIOS:                            LENOVO N27ET36W (1.22 ), 04/07/2019
Répertoire Windows:                         C:\WINDOWS
Répertoire système:                         C:\WINDOWS\system32
Périphérique d’amorçage:                    \Device\HarddiskVolume1
Option régionale du système:                fr;Français (France)
Paramètres régionaux d’entrée:              fr;Français (France)
Fuseau horaire:                             (UTC+01:00) Bruxelles, Copenhague, Madrid, Paris
Mémoire physique totale:                    32 379 Mo
Mémoire physique disponible:                12 884 Mo
Mémoire virtuelle : taille maximale:        57 979 Mo
Mémoire virtuelle : disponible:             30 613 Mo
Mémoire virtuelle : en cours d’utilisation: 27 366 Mo
Emplacements des fichiers d’échange:        C:\pagefile.sys
Domaine:                                    WORKGROUP
Serveur d’ouverture de session:             \\DESKTOP-U2EVGBM
Correctif(s):                               15 Corrections installées.
                                            [01]: KB4601554
                                            [02]: KB4560366
                                            [03]: KB4561600
                                            [04]: KB4562830
                                            [05]: KB4566785
                                            [06]: KB4570334
                                            [07]: KB4577266
                                            [08]: KB4577586
                                            [09]: KB4580325
                                            [10]: KB4586864
                                            [11]: KB4589212
                                            [12]: KB4593175
                                            [13]: KB4598481
                                            [14]: KB5000802
                                            [15]: KB5000858
Carte(s) réseau:                            6 carte(s) réseau installée(s).
                                            [01]: TAP-Windows Adapter V9
                                                  Nom de la connexion : Ethernet 2
                                                  DHCP activé :         Oui
                                                  Serveur DHCP :        10.36.127.254
                                                  Adresse(s) IP
                                                  [01]: 10.36.115.194
                                                  [02]: fe80::d527:58a7:4965:c2ad
                                            [02]: Bluetooth Device (Personal Area Network)
                                                  Nom de la connexion : Connexion réseau Bluetooth
                                                  État :                Support déconnecté
                                            [03]: Intel(R) Dual Band Wireless-AC 8265
                                                  Nom de la connexion : Wi-Fi
                                                  DHCP activé :         Oui
                                                  Serveur DHCP :        192.168.0.254
                                                  Adresse(s) IP
                                                  [01]: 192.168.0.41
                                            [04]: Intel(R) Ethernet Connection (4) I219-LM
                                                  Nom de la connexion : Ethernet
                                                  État :                Support déconnecté
                                            [05]: Hyper-V Virtual Ethernet Adapter
                                                  Nom de la connexion : vEthernet (Default Switch)
                                                  DHCP activé :         Non
                                                  Adresse(s) IP
                                                  [01]: 172.23.192.1
                                                  [02]: fe80::1186:c938:a49b:4bda
                                            [06]: Hyper-V Virtual Ethernet Adapter
                                                  Nom de la connexion : vEthernet (minishift)
                                                  DHCP activé :         Non
                                                  Adresse(s) IP
                                                  [01]: 172.16.0.1
                                                  [02]: fe80::3881:3360:4d27:e878
Configuration requise pour Hyper-V:         Un hyperviseur a été détecté. Les fonctionnalités nécessaires à Hyper-V ne seront pas affichées.
```

### Steps to reproduce

The CRC integration into CodeReady Studio that detects if CRC setup needs to be executed. It is based on the existence of:

1. $HOME/.crc folder
2. $HOME/.crc/crc.json file
3. $HOME/.crc/crc.log file
4. $HOME/.crc/bin folder
5. $HOME/.crc/bin/oc or $HOME/.crc/bin/oc.exe file

if check returns false we run `crc setup` then recheck before running `crc start`. As the oc binary is not extracted under $HOME/.crc/bin start integration is failing.
@robstryker can comment about Linux use case

Is this change permanent ?

### Logs

Before gather the logs try following if that fix your issue
```bash
$ crc delete -f
$ crc cleanup
$ crc setup
$ crc start --log-level debug
```

Please consider posting the output of `crc start --log-level debug`  on http://gist.github.com/ and post the link in the issue.

**Comment by cfergeau** [comment]
> Would you be able to remove that code if we provided a crc setup --dry-run command which would tell you whether any of the crc setup tests failed, but without trying to do any changes on the user's machine? 

The tray actually has a very similar need, but which is currently only being solved through the http API https://github.com/code-ready/crc/pull/2162

**Comment by jeffmaury** [comment]
Yes but we would need json output then

**Comment by jeffmaury** [comment]
Our single point of integration is the crc binary as we need to cover use case where nothing is installed or running on the user's machine

**Comment by gbraad** [comment]
I am not understanding why you would need to check for:

```
$HOME/.crc folder
$HOME/.crc/crc.json file
$HOME/.crc/crc.log file
$HOME/.crc/bin folder
$HOME/.crc/bin/oc or $HOME/.crc/bin/oc.exe file
```

Issuing a `crc status` would at least indicate that the VM exists... which can only exist when the setup has been run. However, a request to provide a 'ready' state would have certainly been accepted.

**Comment by gbraad** [comment]
> crc binary

The executable might not be delivered in future as for macOS and Windows only native installers will be available.

**Comment by praveenkumar** [comment]
This check is only to make sure if studio need to run `crc setup` or not just to avoid the `sudo` prompt (I hope). `crc setup` is intelligent enough to not ask any sudo permission if the required changes are already in place. Shouldn't studio can always run `crc setup` and don't worry about all those changes?

**Comment by guillaumerose** [comment]
Shouldn't crc start always run crc setup checks? We could remove then crc setup.

**Comment by jeffmaury** [comment]
> 
> 
> > Our single point of integration is the crc binary
> 
> The executable might not be delivered in future as for macOS and Windows only native installers will be available.

When is this packaging changed planned ? We need to integrate that into our roadmap as we also provide download and installation (NOP as of yet) into our tooling

**Comment by jeffmaury** [comment]
> 
> 
> This check is only to make sure if studio need to run `crc setup` or not just to avoid the `sudo` prompt (I hope). `crc setup` is intelligent enough to not ask any sudo permission if the required changes are already in place. Shouldn't studio can always run `crc setup` and don't worry about all those changes?

No the check is here to detect crc setup must be run as it's not integrated into crc start. Please note that as we provide crc download from the IDE, the workstation could not have been initialized

**Comment by cfergeau** [comment]
> When is this packaging changed planned ? We need to integrate that into our roadmap as we also provide download and installation (NOP as of yet) into our tooling

We already ship a macOS .pkg with the latest crc releases. It has the benefit of providing signed binaries. This was mentioned in https://github.com/redhat-developer/vscode-openshift-tools/issues/1817#issuecomment-805164029
With this installer, the `crc` binary is available in `/usr/local/bin` though.

**Comment by jeffmaury** [comment]
Yes but I meant breaking change as I understood crc.exe as a starting point would not be available anymore. This is still not the case for MacOS as .tar.xz is still there

**Comment by cfergeau** [comment]
We strongly recommend to switch to the .pkg though, as this provides signed binaries contrary to the tarball, and we may stop shipping the tarball soon.

**Comment by jeffmaury** [comment]
Yes I got it but in order to plan it, we need to know when the old packaging will be removed

**Comment by gbraad** [comment]
Undecided, but will set up a meeting to discuss some of the upcoming 'features' and changes, but also to investigate needs. /cc: @code-ready/crc-team

**Comment by gbraad** [comment]
BTW, we were already looking into this for our own purposes; https://github.com/code-ready/crc/pull/2162 but a definitive solution hasn't been defined. It should likely be a field added to the `crc status` output.

**Comment by robstryker** [comment]
> Shouldn't crc start always run crc setup checks? We could remove then crc setup.

So I just thought I'd take this moment to show you a glimpse from a user:
```
[rob@rawbeast crc-linux-1.24.0-amd64]$ ./crc status -o json
{
  "success": false,
  "error": "Machine does not exist. Use 'crc start' to create it"
}
[rob@rawbeast crc-linux-1.24.0-amd64]$ ./crc start
Run 'crc setup' to unpack the bundle to disk
```

I run crc status, it tells me to run crc start. I run crc start, it tells me to run crc setup. Even if crc status was updated to recognize that it's not set up yet and provide a more clear error message, our tools still aren't going to go parsing specific error messages, as the error messages can change at any time or any locale. So that's not really a possibility. 

Other things that are important to note:  When we run `crc start`, we launch the process through the equivalent of Runtime.exec(). It's wrapped in eclipse APIs, but that's what we do. We then link the output to a console view so the user can see it. 

This is insufficient for `crc status`. For `crc status` we need a PTY terminal (to hide typed passwords) and input ability. Our current hack / workaround is to launch it through a PTY process launcher and connect that to the terminal view in eclipse. Without this workaround, we can't type the root credentials invisibly etc etc. This is something that the cdk / crc team has overlooked often in the past. We generally don't like the terminal view. It doesn't work as well, it's more overhead, it's part of eclipse TM project (which has no real maintainers to speak of) and has been buggy before. Relying on this dependency is dangerous. 

The point here is I don't think we'd be very happy if `crc setup` was removed, and `crc start` did it automatically as part of the first startup. If that happened, we'd need to switch our entire launch to use the terminal view, which would be quite difficult when wrapped with the servertools API. It's fine for us to do that on our own for one-off processes (like crc setup), but when it's part of the "start server" action, well, it becomes way more difficult.

**Comment by gbraad** [comment]
According to the documentation the steps are:

1. `crc setup`
2. `crc start`
3. and usage

While I agree that this is very much sending people for a runaround. Though, `crc start` concludes the right thing.

> For crc status we need a PTY terminal (to hide typed passwords) 

^^^ are you sure? I believe you meant `crc setup`. 

> workaround is to launch it through a PTY process launcher

If so, yes... but this will be handled in 'future' by the OS on macOS as this will be replaced by the yellow package.

> we'd be very happy if crc setup was removed, and crc start did it automatically as part of the first startup

You can prevent the popups and requests for the password IF you supply the pull-secret by file via `crc config set`. This has been implemented IIRC by the VScode adapters. We worked to add configu options which can be provided beforehand to prevent those

In other words, a quick conversation in our IRC channel or slack channel would have been able to resolve this.

**Comment by jeffmaury** [comment]
> 
> 
> According to the documentation the steps are:
> 
>     1. `crc setup`
> 
>     2. `crc start`
> 
>     3. and usage
> 
> 
> While I agree that this is very much sending people for a runaround. Though, `crc start` concludes the right thing.
> 
> > For crc status we need a PTY terminal (to hide typed passwords)
> 
> ^^^ are you sure? I believe you meant `crc setup`.
> 
> > workaround is to launch it through a PTY process launcher
> 
> If so, yes... but this will be handled in 'future' by the OS on macOS as this will be replaced by the yellow package. but ...
> 
> > we'd be very happy if crc setup was removed, and crc start did it automatically as part of the first startup
> 
> You can prevent the popups and requests for the password IF you supply the pull-secret by file via `crc config set`. This has been implemented IIRC by the VScode adapters. We worked to add configu options which can be provided beforehand to prevent those
Are you sure ? We provide pull secret via command line arguments, Rob was referring to sudo like password which I don't think can be prevented
> 
> In other words, a quick conversation in our IRC channel or slack channel would have been able to resolve this.

**Comment by cfergeau** [comment]
> I run crc status, it tells me to run crc start. I run crc start, it tells me to run crc setup. Even if crc status was updated to recognize that it's not set up yet and provide a more clear error message, our tools still aren't going to go parsing specific error messages, as the error messages can change at any time or any locale. So that's not really a possibility.

This is very useful feedback for us, always useful to know the hurdles people encounter in their use of crc. Just reach out to us on slack or file issues as soon as you encounter such issues so that we can work together towards improving crc usability.

**Comment by gbraad** [comment]
Current detection code is here https://github.com/jbosstools/jbosstools-openshift/blob/a56ac0b772569df4b19564fcd067e6de7be1d620/plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/core/adapter/CRC100Server.java#L65-L85

this assumes that `crc.log` exists, that us created for any command that is run.

**Comment by cfergeau** [comment]
Hopefully #2179 would fix this issue?
```
$ ./crc setup --checks-only
INFO Checking if running as non-root
INFO Checking if podman remote executable is cached
INFO Checking if admin-helper executable is cached
admin-helper executable is not cached

$ echo $?
2

$ ./crc setup --checks-only -ojson
INFO Checking if running as non-root
INFO Checking if podman remote executable is cached
INFO Checking if admin-helper executable is cached
{
"success": false,
"error": "admin-helper executable is not cached"
}

$ echo $?
0
```

**Comment by jeffmaury** [comment]
Why do we get non JSON output ? Is it using a different stream ?

**Comment by cfergeau** [comment]
`crc setup` has the same behaviour, `INFO` goes to `stderr`, the json goes to `stdout`.

**Comment by cfergeau** [comment]
@jeffmaury would this be do what you need? or is this better for you if we provide this functionality differently? I'd like to address this in 1.25, and code freeze is at the end of the week.

**Comment by jeffmaury** [comment]
No we should be good assumed the JSON goes to a single stream.

**Comment by cfergeau** [comment]
This is reusing the existing code for json output, so  this should be fine.

**Comment by cfergeau** [comment]
`crc setup --check-only` will be in the 1.25 release, and once you use this, this should fix the issue you were having, feel free to reopen this issue if you think there should be more discussions around this.
